### PR TITLE
New feature: Schema registry

### PR DESCRIPTION
**Pull request by ttossavainen**
https://github.com/FrendsPlatform/Frends.Kafka/issues/6

## Summary by CodeRabbit

- **New Features**
  - Introduced support for Confluent Schema Registry based Avro messaging.
  - Added options for encoding message keys and improved debugging capabilities.
  - Enhanced Kafka consumer to handle both basic string and Avro messages.

- **Bug Fixes**
  - Improved handling of Kafka consumer configurations and connections.

- **Documentation**
  - Updated changelog with significant enhancements and modifications.
  - Clarified property behaviors and added new parameters in documentation.

- **Refactor**
  - Renamed properties for clarity and improved configurability of the Kafka consumer.

- **Tests**
  - Expanded test suites to cover a broader range of consumption scenarios, including edge cases and different security mechanisms.
  - Introduced comprehensive test cases for message consumption functionality.

- **Chores**
  - Updated project dependencies to latest versions for improved performance and security.

### CODE REVIEW

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+using Confluent.SchemaRegistry;
+using Confluent.SchemaRegistry.Serdes;
+using Frends.Kafka.Consume.Definitions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Globalization;
+using System.Runtime.InteropServices;
+using System.Text;
+
+#pragma warning disable CS8600,CS8603,CS8601,CS8604, CS8602
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs:14]: Please comment on why those warnings are and why it is ok to disable those

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+namespace Frends.Kafka.Consume.Tests
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs]: Use file-wise namespace to reduce indents overall.
So just 
`namespace XXX;`
`internal class YYY {}`

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+                Message<string, string> message = new()
+                {
+                    Key = msgKey ?? null,
+                    Value = msg
+                };
+
+                await producer.ProduceAsync(topicPartition, message, cancellationToken);
+            }
+            catch (Exception)
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs]: This does nothing, please remove

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+            if (!string.IsNullOrEmpty(schemaRegistry.SslCaLocation))
+            {
+                schemaRegistryConfig.SslCaLocation = AssignIfNotNullOrEmpty(schemaRegistry.SslCaLocation, string.Empty);
+                schemaRegistryConfig.SslKeystorePassword = AssignIfNotNullOrEmpty(schemaRegistry.SslKeystorePassword, string.Empty);
+                schemaRegistryConfig.SslKeystoreLocation = AssignIfNotNullOrEmpty(schemaRegistry.SslKeystoreLocation, string.Empty);
+            }
+
+            TopicPartition topicPartition = null;
+            if (input.Partition >= 0)
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs]: Is the `else` missing here?..

Comment (jefim) [reply Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/MessageProducer.cs]: Ah, I see that this is done below. Can we do the same logic for this as in ProduceBasic? It is a bit confuding why in ProductBasic it is done one way and here you do another way.

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs [hunk 5]
```diff
@@ -0,0 +1,1167 @@
...
+    private readonly string? _schemaRegistryAPIKeySecret = Environment.GetEnvironmentVariable("ConfluentKafka_SchemaRegistryAPIKeySecret");
+
+    private Input _input = new();
+    private Options _options = new();
+    private Sasl _sasl = new();
+    private Socket _socket = new();
+    private Ssl _ssl = new();
+    private SchemaRegistry _schemaRegistry = new();
+
+    private readonly string _records = @"{
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs]: Can you check why github actions is giving those warnings about _records not used?

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs [hunk 6]
```diff
@@ -0,0 +1,1167 @@
...
+    ""key3"": 3
+  },
+  ""fixedField"": ""YWJjZA=="",
+  ""unionField"": ""Hello, Union!"",
+  ""recordField"": {
+    ""nestedField"": ""Hello, Nested!""
+  }
+}
+";
+    private readonly string _schema = @"{
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs]: Can you check why github actions is giving those warnings about _schema not used?

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs [hunk 7]
```diff
@@ -0,0 +1,1167 @@
...
+            SslKeyPassword = null,
+            SslKeyPem = null,
+            SslKeystoreLocation = null,
+            SslKeystorePassword = null,
+            SslSigalgsList = null,
+        };
+    }
+
+    //[TearDown]
+    //public void Cleanup()
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs:150]: Remove if this is not needed.

Comment (jefim) [reply Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs:150]: Or comment why this is needed :)

File: Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs [hunk 8]
```diff
@@ -19,75 +23,141 @@ public class Kafka
...
+                ? ConsumeAvro(input.Partition, input.Topic, input.MessageCount, input.Timeout, options.EncodeMessageKey, schemaRegistry, Configurations(input, options, socket, sasl, ssl), cancellationToken)
+                : ConsumeBasic(input.Partition, input.Topic, input.MessageCount, input.Timeout, options.EncodeMessageKey, Configurations(input, options, socket, sasl, ssl), cancellationToken);
 
-        TopicPartition topicPartition;
-        if (input.Partition >= 0)
-        {
-            topicPartition = new TopicPartition(input.Topic, new Partition(input.Partition));
         }
-        else
+        catch (Exception)
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs]: This does not do anything :)

File: Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs [hunk 9]
```diff
@@ -19,75 +23,141 @@ public class Kafka
...
+                            ? encodeKey ? Encoding.UTF8.GetString(cr.Message.Key) : cr.Message.Key
+                            : null,
+                        Value = cr.Message.Value ?? null,
                     });
                 else
                     break;
             }
         }
-        catch (Exception ex)
+        catch (Exception)
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs]: Does not do anything

File: Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs [hunk 10]
```diff
@@ -19,75 +23,141 @@ public class Kafka
...
+                        Key = cr.Message.Key != null
+                            ? encodeKey ? Encoding.UTF8.GetString(cr.Message.Key) : cr.Message.Key
+                            : null,
+                        Value = cr.Message.Value ?? null,
+                    });
+                else
+                    break;
+            }
+        }
+        catch (Exception)
```
Comment (jefim) [inline Frends.Kafka.Consume/Frends.Kafka.Consume/Consume.cs]: Does not do anything

**Review by jefim — CHANGES_REQUESTED** [review]

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs (hunk 5 above)
Comment (ttossavainen) [reply Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs]: Forgot to remove _records  and _schema from TestBasic.cs. Those are not needed when running "basic" tests.

File: Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs (hunk 6 above)
Comment (ttossavainen) [reply Frends.Kafka.Consume/Frends.Kafka.Consume.Tests/TestBasic.cs]: Forgot to remove _records  and _schema from TestBasic.cs. Those are not needed when running "basic" tests.

**Review by jefim — APPROVED** [review]
